`AttributeError` for inlined generic functions with py312 pep-695 type-parameter syntax

This regex doesn't match functions that use the [PEP 695](https://peps.python.org/pep-0695/) type parameters syntax:
https://github.com/numba/numba/blob/0.60.0dev0/numba/core/ir.py#L29

In my case, this resulted in:

```
.venv/lib/python3.12/site-packages/numba/core/ir.py:85: in _raw_function_name
    return self._defmatcher.match(defn.strip()).groups()[0]
E   AttributeError: 'NoneType' object has no attribute 'groups'
        defn       = 'def drink_cigar[N: int](X: MatrixF8[_1, N], b: VectorF8[_2]) -> VectorF8[N]:\n'
        self       = Loc(filename=/projects/clowns_are_trains/inflate_bouncy_castle.py, line=192, col=None)
```

This `drink_cigar()` function is wearing a `@njit(inline='always')`.
To give a bit more context, `drink_cigar` is passed as argument to some other (non-inlined) `@njit()` function, `paint_donkeys(drink_cigar, ...)` is this case.
Back when I was still using the archaic `TypeVars`, the code in question was doing what it was supposed to (without errors or warnings).

--- a/docs/upcoming_changes/9459.bug_fix.rst
+++ b/docs/upcoming_changes/9459.bug_fix.rst
@@ -1,0 +1,6 @@
+Allow use of Python 3.12 PEP-695 type parameter syntax
+------------------------------------------------------
+
+A patch is added to properly parse the PEP 695 syntax. While Numba 
+does not yet take advantage of type parameters, it will no longer erroneously 
+reject functions defined with the new Python 3.12 syntax. 

--- a/numba/core/ir.py
+++ b/numba/core/ir.py
@@ -26,7 +26,7 @@ class Loc(object):
     """Source location
 
     """
-    _defmatcher = re.compile(r'def\s+(\w+)\s*\[.*\]\s*\(.*')
+    _defmatcher = re.compile(r'def\s+(\w+)\s*[\[\(]')
 
     def __init__(self, filename, line, col=None, maybe_decorator=False):
         """ Arguments:
@@ -82,10 +82,11 @@ class Loc(object):
     def _raw_function_name(self):
         defn = self._find_definition()
         if defn:
-            return self._defmatcher.match(defn.strip()).groups()[0]
-        else:
-            # Probably exec(<string>) or REPL.
-            return None
+            m = self._defmatcher.match(defn.strip())
+            if m:
+                return m.groups()[0]
+        # Probably exec(<string>) or REPL.
+        return None
 
     def get_lines(self):
         if self.lines is None:

--- a/numba/core/ir.py
+++ b/numba/core/ir.py
@@ -26,7 +26,7 @@ class Loc(object):
     """Source location
 
     """
-    _defmatcher = re.compile(r'def\s+(\w+)\s*[\[\(]')
+    _defmatcher = re.compile(r'def\s+(\w+)')
 
     def __init__(self, filename, line, col=None, maybe_decorator=False):
         """ Arguments:

--- a/numba/core/ir.py
+++ b/numba/core/ir.py
@@ -26,7 +26,7 @@ class Loc(object):
     """Source location
 
     """
-    _defmatcher = re.compile(r'def\s+(\w+)\(.*')
+    _defmatcher = re.compile(r'def\s+(\w+)\s*\[.*\]\s*\(.*')
 
     def __init__(self, filename, line, col=None, maybe_decorator=False):
         """ Arguments:

--- a/numba/tests/test_exceptions.py
+++ b/numba/tests/test_exceptions.py
@@ -3,7 +3,7 @@ import sys
 import traceback
 
 from numba import jit, njit
-from numba.core import types, errors
+from numba.core import types, errors, utils
 from numba.tests.support import (TestCase, expected_failure_py311,
                                  expected_failure_py312,
                                  )
@@ -309,9 +309,18 @@ class TestRaising(TestCase):
         # issue #3428
         simple_raise = "def f(a):\n  raise exc('msg', 10)"
         assert_raise = "def f(a):\n  assert a != 1"
-        for f_text, exc in [(assert_raise, AssertionError),
-                            (simple_raise, UDEArgsToSuper),
-                            (simple_raise, UDENoArgSuper)]:
+        py312_pep695_raise = "def f[T: int](a: T) -> T:\n  assert a != 1"
+        py312_pep695_raise_2 = "def f[T: int\n](a: T) -> T:\n  assert a != 1"
+        test_cases = [
+            (assert_raise, AssertionError),
+            (simple_raise, UDEArgsToSuper),
+            (simple_raise, UDENoArgSuper),
+        ]
+        if utils.PYVERSION >= (3, 12):
+            # Added for https://github.com/numba/numba/issues/9443
+            test_cases.append((py312_pep695_raise, AssertionError))
+            test_cases.append((py312_pep695_raise_2, AssertionError))
+        for f_text, exc in test_cases:
             loc = {}
             exec(f_text, {'exc': exc}, loc)
             pyfunc = loc['f']

